### ISSUE TITLE
Making `@tactic` arguments visible leads to unsolved constraints

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#6781]
This is `test/Succeed/TacticModality.agda`, but with explicit tactic arguments.
```agda
open import Agda.Builtin.Reflection
open import Agda.Builtin.Unit
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

variable
  A : Set

super-tac : Term → TC ⊤
super-tac hole = unify hole (lit (nat 101))

solver : Nat → Term → TC ⊤
solver n hole = unify hole (lit (nat (n + 1)))

-- Changing the tactic argument to visible (over test/Succeed/TacticModality.agda).

foo : (@(tactic super-tac) n : Nat)
      (@(tactic solver n)  x : A) → A
foo  n x = x

number : Nat
number = foo _ _

check : number ≡ 102
check = refl
```
This produces unsolved constraints, in contrast to the variant with hidden tactic arguments.
I think hiding or not should not make such a difference here.

**Comment by nad** [comment]
> This produces unsolved constraints, in contrast to the variant with hidden tactic arguments. I think hiding or not should not make such a difference here.

I suspect that the feature was designed in this way.

What semantics do you have in mind? When is the tactic used? Don't you want it to be possible to override the tactic by giving an argument explicitly (like in "`f _`")?

**Comment by andreasabel** [comment]
> I suspect that the feature was designed in this way.

Note, however, that it does sometimes work with explicit tactic arguments, e.g. here:
https://github.com/agda/agda/blob/6c6b5c929677dd81f89e7ae4952f1b3a64fc0533/test/Succeed/TacticModality.agda#L28-L33

My opinion is that `f {_} a` should behave exactly like `f a`, and in analogy, for a variant `f'` with explicit argument, like `f' _ a`. 

> What semantics do you have in mind?

I have no semantics for this feature in mind, only symmetry criteria, so I can only say what should _not_ happen.

The doc says, at https://github.com/agda/agda/blob/6c6b5c929677dd81f89e7ae4952f1b3a64fc0533/doc/user-manual/language/reflection.lagda.rst#tactic-arguments
> At calls to f, defaultTo true is called on the metavariable inserted for x if it is not given explicitly. 

What exactly is meant by "not given explicitly" is worth a more concrete definition.  I suspect, from my experiments, is that _argument not given explicitly_ is intended to mean _argument is an unsolved meta_, however, the experiments are not free of contradictions.

@UlfNorell , what semantics did you intend?

**Comment by UlfNorell** [comment]
> Note, however, that it does sometimes work with explicit tactic arguments, e.g. here:

It does indeed work as intended here. You can tell because this also works:

```agda
_ : bar _ ≡ 102
_ = refl
```

> My opinion is that `f {_} a` should behave exactly like `f a`

I'm pretty sure we've had this discussion before (for `f {{_}}`). In think the most principled behaviour is that `_` means "insert a metavariable solved by unification", and `{...}` means "give an implicit argument explicitly". Having special rules for when an argument is a syntactic `_` seems like a bad idea.

Following this principle there are three choices
1. Add a warning that explicit tactic arguments are pointless
2. Fail on explicit (and instance) tactic arguments
3. Have `@tactic` imply implicit

I'm leaning towards option 2.

**Comment by andreasabel** [comment]
Welcome back, Ulf!

> > My opinion is that `f {_} a` should behave exactly like `f a`
> 
> I'm pretty sure we've had this discussion before (for `f {{_}}`).

I do remember! 😆 

> 2\. Fail on explicit (and instance) tactic arguments

I agree to this choice.
It is a recoverable error, so it should be an error warning.

So, actually it is

> 1. Add a warning that explicit tactic arguments are pointless

**Comment by andreasabel** [comment]
What about tactics in record fields then, @UlfNorell ? We currently support them even when the fields are not hidden.
- #5712
https://github.com/agda/agda/blob/3b98c183052565193573af8bfa1e83c18a9b5faa/test/Succeed/Issue5712.agda#L8-L19

**Comment by UlfNorell** [comment]
What does "support" mean in this case? Do you have an example where the tactic runs?

**Comment by andreasabel** [comment]
> What does "support" mean in this case? Do you have an example where the tactic runs?

Yes, this one, derived from #5358:
```agda

open import Agda.Builtin.Unit
open import Agda.Builtin.Bool
open import Agda.Builtin.Reflection renaming (bindTC to _>>=_)

defaultTo : {A : Set} (x : A) → Term → TC ⊤
defaultTo x goal = do
  `x ← quoteTC x
  unify goal `x

record Class : Set where
  constructor con
  field
    x : Bool → Bool
    @(tactic defaultTo x) y : Bool → Bool
open Class

testB : Class
testB .x = λ b → b
```

**Comment by andreasabel** [comment]
@UlfNorell In the syntactic approach you favor (rather than a more semantic one like _metavariable with attached tactic_), it seems you cannot curry functions with tactic arguments.
E.g. this fails:
```agda
open import Agda.Builtin.Reflection
open import Agda.Builtin.Unit
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality

super-tac : Term → TC ⊤
super-tac hole = unify hole (lit (nat 101))

solver : Nat → Term → TC ⊤
solver n hole = unify hole (lit (nat (n + 1)))

record Foo : Set where
  field
    @(tactic super-tac) {n} : Nat
    @(tactic solver n)  {x} : Nat

foo : {Foo} → Nat
foo {record{x = x}} = x

number : Nat
number = foo

check : number ≡ 102
check = refl
```

**Comment by UlfNorell** [comment]
> Yes, this one

In that case it makes sense to not given an error, no? We're only suggesting the error because explicit tactic arguments are pointless.

> In the syntactic approach you favor

Just to be clear, this isn't a hypothetical approach that I'm suggesting, *this is how Agda actually works*.

> you cannot curry functions with tactic arguments

Indeed. This is the same behaviour as for instance arguments. I think if you want to suggest a complete overhaul of our handling of metavariables where we aggressively eta expand record metas and run tactics/instance search on corresponding fields, that should maybe be its own issue.

**Comment by andreasabel** [comment]
> In that case it makes sense to not given an error, no? We're only suggesting the error because explicit tactic arguments are pointless.

Not sure, it would be more uniform to then also disallow tactics on non-hidden fields.  This is easier to communicate, and putting hiding annotation on a record field isn't much work.  As I understand it from our discussion, tactics for non-hidden arguments were never intended and it is just an oversight in the implementation that this was not checked for.

**Comment by UlfNorell** [comment]
Sure.

**Comment by andreasabel** [comment]
I started working on the new `UselessTactic` warning on https://github.com/agda/agda/tree/issue-6781, but will finish it only after #6784 which fixes the ranges of tactic attributes so that we can have correct dead code highlighting.

**Comment by jespercockx** [comment]
@andreasabel Since #6784 is fixed now, could your work here be turned into a PR?

**Comment by jespercockx** [comment]
Anyway it's not a super urgent issue, so I'm moving it to `later`.
